### PR TITLE
Add option for `bins` input type - extend data model by entities

### LINKED ISSUE #58: Include bin or taxa (entity) information - extend data model
... to allow strain-wise plotting of binder-nonbinder ratios

### PR DESCRIPTION
**Pull request by skrakau**
- add testdata and profile for `bins` input type (not activated for GitHub actions yet)

## PR checklist

- [x] This comment contains a description of changes (with reason)
- [ ] `CHANGELOG.md` is updated
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] Documentation in `docs` is updated
- [ ] If necessary, also make a PR on the [nf-core/metapep branch on the nf-core/test-datasets repo](https://github.com/nf-core/test-datasets/pull/new/nf-core/metapep)

### CODE REVIEW

**Comment by lkuchenb** [comment]
Still WIP to remodel the old functionality with the new data model

**Comment by skrakau** [comment]
- [x] update_protein_ids -> new name
- [x] ifEmpty() ...
- [x] bin weights: file basename - weight
- [x] bin basename -> tuple val(), ...
- [x] create entity_name from bin_basename
- [ ] update_protein_ids.py: handling for bare proteins still missing
- [x] prodigal on each bin and on each assembly
- [x] predict_proteins: val(bin_basename) - bin base name or false for assembly input

- [x] assign entity weights -> separate process
- [x] check if for all assembley configs weight is given

- [x] handle HTTP input option for `bins` type
- [ ] update CHANGELOG

File: bin/update_protein_ids.py [hunk 1]
```diff
@@ -29,54 +29,131 @@
...
+            # Check validity of runtime arguments
+            if not args.entrez_proteins_assemblies or not args.entrez_proteins_microbiomes:
+                sys.exit("The --entrez-* flags have to be specified together")
+            # Read proteins and associations
+            proteins = pd.read_csv(args.entrez_proteins, sep='\t')
+            proteins_assemblies = pd.read_csv(args.entrez_proteins_assemblies, '\t')\
+                    .rename(columns={
+                        'assembly_id' : 'entity_name',
+                        })
+            proteins_microbiomes   = pd.read_csv(args.entrez_proteins_microbiomes, '\t')
```
Comment (skrakau) [inline bin/update_protein_ids.py]: ok, it is somehow confusing that we get rid rid of the `proteins_microbiomes` table, but still use this as an intermediate file. Could we adjust this somehow to the new format? e.g. store `microbiome_entities_tmp` using `entity_name`, i.e. `assembly_id` and then here just update it with `entity_id`?

File: bin/update_protein_ids.py [hunk 2]
```diff
@@ -29,54 +29,131 @@
...
+
+            # Assign entity_id
+            entities = proteins_assemblies[['entity_name']].drop_duplicates()
+            entities['entity_id'] = range(next_entity_id, next_entity_id + len(entities))
+            next_entity_id += len(entities)
+
+            entities_microbiomes_proteins = proteins_assemblies\
+                    .merge(proteins)\
+                    .merge(entities)\
+                    .merge(proteins_microbiomes)[['entity_id', 'protein_id', 'microbiome_id']].drop_duplicates()
```
Comment (skrakau) [inline bin/update_protein_ids.py]: where would duplicates come from? there shouldn't be any, no?

File: main.nf [hunk 3]
```diff
@@ -310,74 +390,129 @@ process predict_proteins {
...
+    val microbiome_id into ch_pred_proteins_microbiome_ids                  // Emit microbiome ID
+    val bin_basename into ch_pred_proteins_bin_basename
+    file("proteins.pred_${microbiome_id}*.tsv.gz") into ch_pred_proteins     // Emit protein tsv
+    file "coords.pred_${microbiome_id}*.gff"
 
     script:
     def mode = params.prodigal_mode
+    def name = bin_basename ? "${microbiome_id}.${bin_basename}" : "${microbiome_id}"
     """
     gzip -c -d $microbiome_file | prodigal \
```
Comment (lkuchenb) [inline main.nf]: This process cannot handle plain FASTA files, but they are supported as input

Comment (skrakau) [reply main.nf]: Removed plain FASTA format from list of supported file extensions for now

File: bin/assign_entity_weights.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+####################################################################################################
+# for each strain: select largest assembly (for now)
```
Comment (skrakau) [inline bin/assign_entity_weights.py]:
```suggestion
```

File: bin/assign_entity_weights.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+dfs = []
+for mb_id, w_path in zip(args.microbiome_ids, args.weights_files):
+    data = pd.read_csv(w_path, sep='\t').rename(columns={
+        'contig_name' : 'entity_name',
+        'bin_basename' : 'entity_name'
+        })
+    data['microbiome_id'] = mb_id
+    dfs.append(data)
+
+column_names = ["entity_name", "microbiome_id", "weight"]
```
Comment (skrakau) [inline bin/assign_entity_weights.py]: should this not be
```suggestion
column_names = ["entity_name", "microbiome_id", "entity_weight"]
```
?

Comment (lkuchenb) [reply bin/assign_entity_weights.py]: The switch from the user interface naming (`weight`) to the internal data model naming (`entity_weight`) happens in `finalize_microbiome_entities.py` where all weigths are consolidated.

File: bin/finalize_microbiome_entities.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        pass
+    else:
+        raise PartialWeightsError(subset['microbiome_id'].iloc[0])
+    return subset
+
+####################################################################################################
+
+args = parse_args()
+
+if not args.entrez_microbiomes_entities and not args.nucl_microbiomes_entities:
```
Comment (skrakau) [inline bin/finalize_microbiome_entities.py:62]: isn't that the case if assemblies are provided without weights?

Comment (skrakau) [reply bin/finalize_microbiome_entities.py:62]: will be an empty file then, containing only the header

File: main.nf [hunk 7]
```diff
@@ -218,48 +216,128 @@ process create_db_tables {
...
+        bin_basenames : bin_files.collect{ it.name - fasta_suffix }
+    }.set { ch_microbiomes_bins_folders }
+
+// BINS - LOCAL OR REMOTE ARCHIVES
+ch_microbiomes_bins.archives
+    .multiMap { row ->
+        ids : row.microbiome_id
+        files: row.microbiome_path
+    }
+    .set{ch_microbiomes_archives}
```
Comment (skrakau) [inline main.nf:293]:
```suggestion
    .set{ch_microbiomes_bins_archives_packed}
```

File: main.nf [hunk 8]
```diff
@@ -218,48 +216,128 @@ process create_db_tables {
...
+// BINS
+ch_microbiomes_branch.bins
+    .branch {
+            row ->
+            folders : row.microbiome_path.isDirectory()
+            archives : row.microbiome_path.isFile()
+            other: true
+        }
+    .set{ ch_microbiomes_bins }
+
```
Comment (skrakau) [inline main.nf:274]: Very nice restructured!

File: bin/prepare_entity_binding_ratios.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+
+def call_binder(score, method):
+    if method == "syfpeithi":
+        if score >= 0.50:
+            return True
+    else:
+        if score <= 500:
+            return True
+    return False
```
Comment (lkuchenb) [inline bin/prepare_entity_binding_ratios.py]:
```suggestion
        return score >= 0.50:
    else:
        return score <= 500:
```

**Review by lkuchenb — APPROVED** [review]

**Pull request merged**